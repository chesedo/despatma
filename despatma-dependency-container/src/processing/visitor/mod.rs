use despatma_visitor::visitor_mut;

use super::{ChildDependency, Container, Dependency};

pub use add_wildcard_lifetime::AddWildcardLifetime;
pub use extract_async::ExtractAsync;
pub use extract_box_type::ExtractBoxType;
pub use extract_lifetime::ExtractLifetime;
pub use impl_trait_but_registered_concrete::ImplTraitButRegisteredConcrete;
pub use impl_trait_fields::ImplTraitFields;
pub use link_dependencies::LinkDependencies;
pub use owning_managed_dependency::OwningManagedDependency;
pub use replace_impl_generics_with_concrete::ReplaceImplGenericsWithConcrete;
pub use unsupported_registered_types::UnsupportedRegisteredTypes;
pub use wrap_box_type::WrapBoxType;

mod add_wildcard_lifetime;
mod extract_async;
mod extract_box_type;
mod extract_lifetime;
mod impl_trait_but_registered_concrete;
mod impl_trait_fields;
mod link_dependencies;
mod owning_managed_dependency;
mod replace_impl_generics_with_concrete;
mod unsupported_registered_types;
mod wrap_box_type;

visitor_mut!(
    #[helper_tmpl = {
        for dependency in container.dependencies.iter() {
            visitor.visit_dependency_mut(&mut dependency.borrow_mut());
        }
    }]
    Container,
    #[helper_tmpl = {
        for dependency in dependency.dependencies.iter_mut() {
            visitor.visit_child_dependency_mut(dependency);
        }
    }]
    Dependency,
    #[helper_tmpl = {
        visitor.visit_dependency_mut(&mut child_dependency.inner.borrow_mut());
    }]
    ChildDependency,
);

/// A visitor used for validation
/// If the visitor found any errors then they should be emit in [emit_errors].
pub trait ErrorVisitorMut: VisitorMut + Sized {
    fn new() -> Self;

    fn emit_errors(self) {}
}
