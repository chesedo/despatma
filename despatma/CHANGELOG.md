# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1](https://github.com/chesedo/despatma/compare/despatma-v0.1.0...despatma-v0.1.1) - 2024-08-09

### Other
- workspace versioning ([#4](https://github.com/chesedo/despatma/pull/4))
- Update dependencies ([#2](https://github.com/chesedo/despatma/pull/2))
- update failing tests
- force version downgrades
