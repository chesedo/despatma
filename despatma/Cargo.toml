[package]
name = "despatma"
version = "0.1.0"
authors = ["chesedo <pieter@chesedo.me>"]
edition = "2018"
description = "Design Pattern Macro like Loki"
repository = "https://github.com/chesedo/despatma"
license = "MIT"
keywords = ["macro", "design", "patterns"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
despatma-lib = { path = "../despatma-lib", version = "0.1.0" }

proc-macro2 = "1.0"
quote = "1.0"
syn = "1.0"
tokenstream2-tmpl = "=0.1.1"


[dev-dependencies]
despatma-test-helpers = { path = "../despatma-test-helpers" }
despatma-lib = { path = "../despatma-lib", features = ["extra-traits"] }
macrotest = "1.0"
pretty_assertions = "0.6"
syn = { version = "1.0", features = ["extra-traits"] }
trybuild = "1.0"
