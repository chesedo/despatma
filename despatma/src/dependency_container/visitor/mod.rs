use super::{ChildDependency, Container, Dependency};

mod async_visitor;
mod impl_trait_but_registered_concrete;
mod wiring_visitor;

pub use async_visitor::AsyncVisitor;
pub use impl_trait_but_registered_concrete::ImplTraitButRegisteredConcrete;
pub use wiring_visitor::WiringVisitor;

/// A visitor used to validate the struct that will be turned into a dependency container.
/// If the visitor found any errors then they should be emit in [emit_errors].
pub trait Visit {
    /// Visit the top level container which will be turned into a struct
    fn visit_container(&mut self, container: &Container) {
        visit_container(self, container);
    }

    /// Visit a dependency that was registered
    fn visit_dependency(&mut self, dependency: &Dependency) {
        visit_dependency(self, dependency);
    }

    /// Visit a dependency requested by a registered dependency
    fn visit_child_dependency(&mut self, child_dependency: &ChildDependency) {
        visit_child_dependency(self, child_dependency);
    }

    /// Emit any errors that were found during the visit
    fn emit_errors(self);
}

fn visit_container<V: Visit + ?Sized>(visitor: &mut V, container: &Container) {
    for dependency in container.dependencies.values() {
        visitor.visit_dependency(&dependency.borrow());
    }
}

fn visit_dependency<V: Visit + ?Sized>(visitor: &mut V, dependency: &Dependency) {
    for child_dependency in &dependency.dependencies {
        visitor.visit_child_dependency(child_dependency);
    }
}

fn visit_child_dependency<V: Visit + ?Sized>(
    _visitor: &mut V,
    _child_dependency: &ChildDependency,
) {
}

/// A mutable visitor used to update any dependencies or their children
pub trait VisitMut {
    fn visit_container_mut(&mut self, container: &mut Container) {
        visit_container_mut(self, container);
    }

    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        visit_dependency_mut(self, dependency);
    }

    fn visit_child_dependency_mut(&mut self, child_dependency: &mut ChildDependency) {
        visit_child_dependency_mut(self, child_dependency);
    }
}

fn visit_container_mut<V: VisitMut + ?Sized>(visitor: &mut V, container: &mut Container) {
    for dependency in container.dependencies.values_mut() {
        visitor.visit_dependency_mut(&mut dependency.borrow_mut());
    }
}

fn visit_dependency_mut<V: VisitMut + ?Sized>(visitor: &mut V, dependency: &mut Dependency) {
    for child_dependency in &mut dependency.dependencies {
        visitor.visit_child_dependency_mut(child_dependency);
    }
}

fn visit_child_dependency_mut<V: VisitMut + ?Sized>(
    _visitor: &mut V,
    _child_dependency: &mut ChildDependency,
) {
}
