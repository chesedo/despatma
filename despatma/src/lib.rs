mod abstract_factory;
mod visitor;

extern crate proc_macro;

use proc_macro::TokenStream;
use syn::punctuated::Punctuated;
use syn::{parse_macro_input, ItemTrait, Token};
use tokenstream2_tmpl::Interpolate;

use abstract_factory::AbstractFactoryAttribute;
use despatma_lib::TraitSpecifier;
use visitor::VisitorFunction;

#[proc_macro_attribute]
pub fn abstract_factory(tokens: TokenStream, trait_expr: TokenStream) -> TokenStream {
    let mut input = parse_macro_input!(trait_expr as ItemTrait);
    let attributes = parse_macro_input!(tokens as AbstractFactoryAttribute);

    attributes.expand(&mut input).into()
}

#[proc_macro_attribute]
pub fn interpolate_traits(tokens: TokenStream, concrete_impl: TokenStream) -> TokenStream {
    let attributes =
        parse_macro_input!(tokens with Punctuated::<TraitSpecifier, Token![,]>::parse_terminated);

    attributes.interpolate(concrete_impl.into()).into()
}

#[proc_macro]
pub fn visitor(tokens: TokenStream) -> TokenStream {
    let input = parse_macro_input!(tokens as VisitorFunction);

    input.expand().into()
}
