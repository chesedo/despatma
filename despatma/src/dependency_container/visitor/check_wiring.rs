use proc_macro_error::emit_error;
use strsim::levenshtein;
use syn::Ident;

use crate::dependency_container::ChildDependency;

use super::{visit_child_dependency, Visit};

/// This visitor is responsible for checking if all dependencies have been registered in the container.
/// So if `a` has a dependency on `b`, this visitor will check if `b` has been registered in the container.
/// If not, it will emit an error.
pub struct CheckWiring {
    dependencies: Vec<Ident>,
    errors: Vec<Error>,
}

#[cfg_attr(test, derive(Eq, PartialEq, Debug))]
struct Error {
    requested: Ident,
    best_match: Option<Ident>,
}

impl CheckWiring {
    /// Create a new `WiringVisitor` with the given available dependencies.
    /// Ie. if this visitor finds a requested child dependency which is not in the list of available dependencies,
    /// it will emit an error.
    pub fn new(dependencies: Vec<Ident>) -> Self {
        Self {
            dependencies,
            errors: Vec::new(),
        }
    }
}

impl Visit for CheckWiring {
    fn visit_child_dependency(&mut self, child_dependency: &ChildDependency) {
        if !self.dependencies.contains(&child_dependency.ident) {
            let best_match =
                get_best_dependency_match(&self.dependencies, &child_dependency.ident.to_string());

            self.errors.push(Error {
                requested: child_dependency.ident.clone(),
                best_match,
            });
        }

        // Keep traversing the tree
        visit_child_dependency(self, child_dependency);
    }

    fn emit_errors(self) {
        let Self { errors, .. } = self;

        for error in errors {
            if let Some(best_match) = error.best_match {
                emit_error!(error.requested, "The '{}' dependency has not been registered", error.requested; hint = best_match.span() => format!("Did you mean `{}`?", best_match));
            } else {
                emit_error!(
                    error.requested,
                    "Dependency not found. Did you forget to add it?";
                    hint = "Try adding it with `fn {}(&self) ...`", error.requested
                );
            }
        }
    }
}

/// The maximum distance between two strings for them to be considered a misspelling.
const MISSPELLING_THRESHOLD: usize = 3;

fn get_best_dependency_match(dependencies: &[Ident], needle: &str) -> Option<Ident> {
    dependencies
        .iter()
        .map(|d| (d, levenshtein(needle, &d.to_string())))
        .filter(|(_, distance)| *distance <= MISSPELLING_THRESHOLD)
        .min_by_key(|(_, distance)| *distance)
        .map(|(d, _)| d.clone())
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use crate::dependency_container::Container;

    use super::*;

    #[test]
    fn check_wiring() {
        let mut wiring_visitor = CheckWiring::new(vec![
            Ident::new("foo", proc_macro2::Span::call_site()),
            Ident::new("bar", proc_macro2::Span::call_site()),
            Ident::new("baz", proc_macro2::Span::call_site()),
        ]);
        let container = Container::from_item_impl(parse_quote!(
            impl Dependencies {
                fn service(&self, zoo: Foo, barrier: Barrier, baz: Baz) -> Service {
                    Service
                }
            }
        ));

        wiring_visitor.visit_container(&container);

        let CheckWiring { errors, .. } = wiring_visitor;

        assert_eq!(
            errors,
            vec![
                Error {
                    requested: Ident::new("zoo", proc_macro2::Span::call_site()),
                    best_match: Some(Ident::new("foo", proc_macro2::Span::call_site())),
                },
                Error {
                    requested: Ident::new("barrier", proc_macro2::Span::call_site()),
                    best_match: None,
                }
            ]
        );
    }
}
