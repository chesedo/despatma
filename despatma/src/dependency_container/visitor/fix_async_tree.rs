use std::{cell::RefCell, collections::HashSet, rc::Rc};

use indexmap::IndexMap;
use syn::Ident;

use crate::dependency_container::Dependency;

use super::VisitMut;

/// Visitor used to determine if any child dependencies in the calltree is async. Because if any
/// child dependencies are async, then the parent dependency must also be async. Which this
/// visitor updates correcly.
pub struct FixAsyncTree {
    dependencies: IndexMap<Ident, Rc<RefCell<Dependency>>>,
    visited_dependencies: HashSet<Ident>,
}

impl FixAsyncTree {
    pub fn new(dependencies: IndexMap<Ident, Rc<RefCell<Dependency>>>) -> Self {
        Self {
            dependencies,
            visited_dependencies: HashSet::new(),
        }
    }
}

impl VisitMut for FixAsyncTree {
    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        // We want to be efficient and not visit the same dependency multiple times
        if self.visited_dependencies.contains(&dependency.sig.ident) {
            return;
        }

        let dependencies: Vec<_> = dependency
            .dependencies
            .iter()
            .filter_map(|d| self.dependencies.get(&d.ident))
            .cloned()
            .collect();

        let has_sync_child = dependencies.iter().any(|d| {
            let mut d = d.borrow_mut();

            // Make sure we update any child dependencies first as they might be in an async calltree.
            // Basically the idea is to visit the calltree in a depth-first manner.
            self.visit_dependency_mut(&mut d);

            d.is_async
        });

        dependency.is_async |= has_sync_child;

        self.visited_dependencies
            .insert(dependency.sig.ident.clone());
    }
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use crate::dependency_container::Container;

    use super::*;

    #[test]
    fn fix_async_tree() {
        let mut container = Container::from_item_impl(parse_quote!(
            impl Dependencies {
                fn config(&self) -> Config {
                    Config
                }

                fn service(&self, config: Config, db: Db) -> Service {
                    Service(config, db)
                }

                async fn logger(&self) -> Logger {
                    Logger
                }

                fn db(&self, logger: Logger) -> Db {
                    Db
                }
            }
        ));

        let mut async_visitor = FixAsyncTree::new(container.dependencies.clone());
        async_visitor.visit_container_mut(&mut container);

        assert!(
            !container
                .dependencies
                .get::<Ident>(&parse_quote!(config))
                .unwrap()
                .borrow()
                .is_async
        );
        assert!(
            container
                .dependencies
                .get::<Ident>(&parse_quote!(logger))
                .unwrap()
                .borrow()
                .is_async
        );
        assert!(
            container
                .dependencies
                .get::<Ident>(&parse_quote!(db))
                .unwrap()
                .borrow()
                .is_async
        );
        assert!(
            container
                .dependencies
                .get::<Ident>(&parse_quote!(service))
                .unwrap()
                .borrow()
                .is_async
        );
    }
}
