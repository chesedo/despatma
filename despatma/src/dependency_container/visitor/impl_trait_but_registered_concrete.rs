use std::{cell::RefCell, rc::Rc};

use indexmap::IndexMap;
use proc_macro_error::emit_error;
use quote::ToTokens;
use syn::{FnArg, Ident, Pat, PatType, ReturnType, Type, TypeImplTrait};

use crate::dependency_container::Dependency;

use super::Visit;

/// Visitor to find any requested dependencies of a concrete type, while the registered dependency
/// returns an `impl Trait`.
pub struct ImplTraitButRegisteredConcrete {
    dependencies: IndexMap<Ident, Rc<RefCell<Dependency>>>,
    errors: Vec<Error>,
}

#[cfg_attr(test, derive(Eq, PartialEq, Debug))]
struct Error {
    requested: PatType,
    registered: TypeImplTrait,
}

impl ImplTraitButRegisteredConcrete {
    pub fn new(dependencies: IndexMap<Ident, Rc<RefCell<Dependency>>>) -> Self {
        Self {
            dependencies,
            errors: Vec::new(),
        }
    }
}

impl Visit for ImplTraitButRegisteredConcrete {
    fn visit_dependency(&mut self, dependency: &Dependency) {
        let path_inputs = dependency
            .sig
            .inputs
            .iter()
            .filter_map(|input| match input {
                FnArg::Receiver(_) => None,
                FnArg::Typed(pat_type) => match pat_type.ty.as_ref() {
                    Type::Path(_) => Some(pat_type),
                    _ => None,
                },
            });

        for pat_type in path_inputs {
            let child_ident = match pat_type.pat.as_ref() {
                Pat::Ident(pat_ident) => pat_ident.ident.clone(),
                _ => continue,
            };

            let Some(child_dependency) = self.dependencies.get(&child_ident) else {
                continue;
            };

            let child_dependency = child_dependency.borrow();
            let child_return_type = match &child_dependency.sig.output {
                ReturnType::Default => continue,
                ReturnType::Type(_, ty) => match ty.as_ref() {
                    Type::ImplTrait(impl_trait) => impl_trait,
                    _ => continue,
                },
            };

            self.errors.push(Error {
                requested: pat_type.clone(),
                registered: child_return_type.clone(),
            });
        }
    }

    fn emit_errors(self) {
        let errors = self.errors;

        for Error {
            requested,
            registered,
        } in errors
        {
            emit_error!(
                requested,
                "Requested type is a concrete type, but the registered type is: `{}`",
                registered.to_token_stream();
                hint = "change this to `{}: {}`", requested.pat.to_token_stream(), registered.to_token_stream()
            )
        }
    }
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use crate::dependency_container::Container;

    use super::*;

    #[test]
    fn impl_trait_but_registered_concrete() {
        let container = Container::from_item_impl(parse_quote!(
            impl Dependencies {
                fn db(&self) -> impl DB {
                    Sqlite
                }

                fn service(&self, db: Sqlite) -> Service {
                    Service(db)
                }
            }
        ));

        let mut visitor = ImplTraitButRegisteredConcrete::new(container.dependencies.clone());
        visitor.visit_container(&container);

        let ImplTraitButRegisteredConcrete { errors, .. } = visitor;

        assert_eq!(
            errors,
            vec![Error {
                requested: parse_quote!(db: Sqlite),
                registered: parse_quote!(impl DB),
            }]
        );
    }
}
