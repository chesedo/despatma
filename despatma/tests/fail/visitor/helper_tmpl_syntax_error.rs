/// Test a syntax error in the helper_tmpl option
mod lib;

use despatma::visitor;
use lib::shapes::{Arc, Circle, Cube};

visitor!(
    #[no_default]
    dyn Circle,

    Arc,

    #[helper_tmpl = {visitor.visit_rectangle(cube.get_back())}]
    dyn Cube,
);

fn main() {}
