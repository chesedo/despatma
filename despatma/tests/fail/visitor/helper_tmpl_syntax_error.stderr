error[E0599]: no method named `visit_rectangle` found for mutable reference `&mut V` in the current scope
  --> $DIR/helper_tmpl_syntax_error.rs:13:30
   |
13 |     #[helper_tmpl = {visitor.visit_rectangle(cube.get_back())}]
   |                              ^^^^^^^^^^^^^^^ method not found in `&mut V`

error[E0599]: no method named `get_back` found for reference `&dyn lib::shapes::Cube` in the current scope
  --> $DIR/helper_tmpl_syntax_error.rs:13:51
   |
13 |     #[helper_tmpl = {visitor.visit_rectangle(cube.get_back())}]
   |                                                   ^^^^^^^^ method not found in `&dyn lib::shapes::Cube`
